src/sqlfluff/core/parser/segments/base.py

```diff
@@ -131,13 +117,21 @@ def __init__(self, segments, pos_marker=None, validate=True):
                 )
 
     def __eq__(self, other):
-        # Equal if type, content and pos are the same
         # NB: this should also work for RawSegment
         return (
             # Same class NAME. (could be constructed elsewhere)
             self.__class__.__name__ == other.__class__.__name__
             and (self.raw == other.raw)
-            and (self.pos_marker == other.pos_marker)
+            # Both must have a non-null position marker to compare.
+            and self.pos_marker
+            and other.pos_marker
```

Comment:
Is there potential for confusing behavior now or someday if one has a `pos_marker` and the other doesn't? (Wondering if, for example, an exception should be raised rather than simply returning `False`?)

Reply:
Potentially but unlikely. `null` position markers are assigned positions almost straight after addition in the fix routines. I've been thinking of the missing position marker more as _delegating_ the positioning to the parent segment, rather than trying to work it out itself.